ZSA assets swap

ConstanceBeguier opened the pull request:
This PR introduces the ability to create `ActionGroup` and `SwapBundle` for performing asset swaps.

This PR modifies the `Builder`:
- A new `reference_notes` element has been added, which contains reference split notes. These notes can be used when creating actions.
- A new `build_action_group` function allows the creation of an `ActionGroup` from a `Builder`. Like when creating a `Bundle`, the `ActionGroup` will initially be without proof and signatures. You will need to call `create_proof` and `apply_signatures` to add them.

This PR creates a new `SwapBundle` structure. A `SwapBundle` is created from a list of `ActionGroups`. During this creation, the following operations are performed
- The evaluation of the value balance of the `SwapBundle` (by summing the value balances of each `ActionGroup`).
- The evaluation of the binding signature key `bsk` of the `SwapBundle` (by summing the `bsk` of each `ActionGroup`).
- The binding signature of the `SwapBundle` which is equal to the signature of the SwapBundle SIGHASH with the binding signature key `bsk`.

TODO
- **Should remove timelimit inside ActionGroup struct (this value is already present in Bundle)**
- In this PR, burning is not allowed in an ActionGroup, but we would like to allow it and store burning into inside ActionGroup.
   - The SpecificBuilderParams struct, the BurnNotEmptyInActionGroup error, and the bundle method need to be updated accordingly.
- A Bundle can now contain an ActionGroup. This modification must be implemented.
```rust
/// An action group.
#[derive(Debug)]
pub struct ActionGroup<A: Authorization, V> {
    /// The action group main content.
    action_group: Bundle<A, V, OrchardZSA>,
    /// The action group timelimit.
    timelimit: u32,
    /// The binding signature key for the action group.
    ///
    /// During the building of the action group, this key is not set.
    /// Once the action group is finalized (it contains a spend authorizing signature for each
    /// action and a proof), the key is set.
    bsk: Option<redpallas::SigningKey<Binding>>,
}
```
- There are currently two functions for computing orchard_action_groups_digest. These should be merged into a single function for better maintainability.
- The function binding_validating_key into swap_bundle.rs must take into account the burns.

ZIP
- The value balance of a swap order is missing in the current ZIP. The matcher (the entity who creates the swap bundle from the swap orders) must know the value balance of each swap order in order to compute the value balance of the swap bundle. The value balance of the swap bundle is equal to the sum of the value balance of each swap order.

ConstanceBeguier committed:
Add split_notes HashMap into Builder

ConstanceBeguier committed:
Add timelimit into Builder and Bundle

ConstanceBeguier committed:
Timelimit is u32 and not u64

ConstanceBeguier committed:
Add sighash evaluation for action groups

ConstanceBeguier committed:
Add ActionGroup tests

ConstanceBeguier committed:
Add matcher action group for fees in tests

ConstanceBeguier committed:
Add SwapBundle struct

ConstanceBeguier committed:
Add SwapBundle functionalities

ConstanceBeguier committed:
Fix errors

ConstanceBeguier committed:
Fix fees

ConstanceBeguier committed:
Fix errors from zsa1 rebasing

ConstanceBeguier committed:
Move new functionalities and structs into swap_bundle.rs file

ConstanceBeguier committed:
Create Actiongroup structure

ConstanceBeguier committed:
Remove timelimit from Builder and Bundle

ConstanceBeguier committed:
Move ActionGroupAuthorized into swap_bundle.rs file

ConstanceBeguier committed:
When building a SwapBundle, remove bsk for each action group

ConstanceBeguier committed:
Add and update comments

ConstanceBeguier committed:
Update hash_action_group and hash_swap_bundle

ConstanceBeguier committed:
Replace split_notes by reference_notes

ConstanceBeguier committed:
Create SpecificBuilderParams

ConstanceBeguier committed:
Update SwapBundle creation

ConstanceBeguier committed:
Add miner fees in swap tests

ConstanceBeguier committed:
Create reference keys in swap tests

All set of keys (user1, user2, matcher, reference) must be different.

ConstanceBeguier committed:
Add some swap tests

ConstanceBeguier committed:
Rename sighsh by action_group_digest

ConstanceBeguier committed:
Add ReferenceNotesAndKeys structure

ConstanceBeguier committed:
Remove allow clippy too_many_arguments on build_and_verify_action_group

ConstanceBeguier committed:
Add reference notes into issuance.rs

ConstanceBeguier committed:
Fix zsa tests due to new IssueBundle format

ConstanceBeguier committed:
Add recipient at TestOutputInfo

alexeykoren commented on src/builder.rs:
```diff
@@ -658,9 +682,34 @@ impl Builder {
+            SpecificBuilderParams::ActionGroupParams(self.reference_notes),
+        )?
+        .unwrap()
+        .0;
```
Actually, we need metadata as well, like in normal build

ConstanceBeguier replied:
Done

ConstanceBeguier committed:
[WIP] Update comments

ConstanceBeguier commented on src/swap_bundle.rs:
```diff
@@ -0,0 +1,239 @@
+            .iter()
+            .fold(
+                self.action_group
+                    .prepare_for_action_group(&mut rng, action_group_digest),
```
It is exactly similar to the prepare function for Bundle.
It is inside the prepare_for_action_group function that you save the digest into self.

ConstanceBeguier committed:
When creating an action group from a builder, return metadata

alexeykoren committed:
Zsa Swap Utils  (#141)

This PR contains a set of additions/changes that allow ZSA Swaps to be
implemented in librustzcash:
- Adds Swap BundleType and Flag
- Adds AuthorizedWithProof trait that defines proof() method for both
Authorized and ActionGroupAuthorized
- Adds utility methods e.g. from_parts() for SwapBundle, is_empty() in
Builder
- Changes visibility of some methods e.g. dummy() for testing in
librustzcash
- Removes ActionGroup structure

alexeykoren committed:
Add swaps flag serialization

alexeykoren committed:
Remove BurnNotEmptyInActionGroup error

alexeykoren committed:
Add burn to action group bundle

alexeykoren committed:
Post-merge fixes

alexeykoren committed:
Set split_flag to true for reference notes

alexeykoren committed:
Use correct split note in test

alexeykoren committed:
After-merge fixes

alexeykoren committed:
Use versioned sig in swap bundle

Commit:
Add burn to bvk

alexeykoren committed:
Merge branch 'zsa1' into zsa_swap_abd6ecc

# Conflicts:
#	src/builder.rs
#	src/pczt/tx_extractor.rs

alexeykoren committed:
Merge zsa1

Commit:
Fmt

Commit:
Clippy

alexeykoren committed:
Merge branch 'zsa1' into zsa_swap

alexeykoren committed:
Add circuit cfg guards

alexeykoren committed:
Add circuit cfg guards

alexeykoren committed:
Update blake

vivek-arte committed:
Updating README (#248)

This adds a Swaps specific disclaimer to the README to highlight it is
for testing use only, and a work-in-progress.
